python/python/bystro/supervised_ppca/gf_dropout_pt.py

```diff
@@ -0,0 +1,299 @@
+            The data
+
+        y : np.array-like,(n_samples,n_prediction)
+            Covariates we wish to predict. For now lazy and assuming
+            logistic regression.
+
+        task : string,default='classification'
+            Is this prediction, multinomial regression, or classification
+
+        progress_bar : bool,default=True
+            Whether to print the progress bar to monitor time
+
+        seed : int,default=2021
+            The random number generator seed used to ensure reproducibility
+
+        Returns
+        -------
+        self : object
+            The model
+        """
+        self._test_inputs(X, y)
+        rng = np.random.default_rng(int(seed))
+        training_options = self.training_options
+        N, p = X.shape
+        self.p = p
+
+        W_, sigmal_, B_ = self._initialize_variables(X)
+        X, y = self._transform_training_data(X, 1.0*y)
+
+        if task == "classification":
+            sigm = nn.Sigmoid()
+            supervision_loss = nn.BCELoss()
+        elif task == "regression":
+            supervision_loss = nn.MSELoss()
+        else:
+            supervision_loss = nn.BCELoss()
+
+        trainable_variables = [W_, sigmal_, B_]
+
+        optimizer = torch.optim.SGD(
+            trainable_variables,
+            lr=training_options["learning_rate"],
+            momentum=training_options["momentum"],
+        )
+        eye = torch.tensor(np.eye(p))
+        one_s = torch.tensor(np.ones(self.n_supervised))
+        softplus = nn.Softplus()
+
+        _prior = self._create_prior()
+
+        myrange = trange if progress_bar else range
+
+        for i in myrange(training_options["n_iterations"]):
+            idx = rng.choice(
+                X.shape[0], size=training_options["batch_size"], replace=False
+            )
+            X_batch = X[idx]
+            y_batch = y[idx]
+
+            sigma = softplus(sigmal_)
+            WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
+            Sigma = WWT + sigma * eye
+
+            like_prior = _prior(trainable_variables)
+
+            # Generative likelihood
+            m = MultivariateNormal(torch.zeros(p), Sigma)
+            like_gen = torch.mean(m.log_prob(X_batch))
+
+            # Predictive lower bound
+            P_x, Cov = _get_projection_matrix(W_, sigma)
+            mean_z = torch.matmul(X_batch, torch.transpose(P_x, 0, 1))
+            eps = torch.rand_like(mean_z)
+            C1_2 = torch.cholesky(Cov)
+            z_samples = mean_z + torch.matmul(eps, C1_2)
+
+            y_hat = (
+                self.delta
+                * torch.matmul(z_samples[:, : self.n_supervised], one_s)
+                + B_
+            )
+            if task == "regression":
+                loss_y = supervision_loss(y_hat, y_batch)
+            else:
+                loss_y = supervision_loss(sigm(y_hat), y_batch)
+
+            WTW = torch.matmul(W_, torch.transpose(W_, 0, 1))
```

Comment:
Ah right, so it depends on whether W\in R^{Lxp} or R^{pxL}, and in the code implementation I match sklearn which is the former.

Reply:
I agree with defining `W` that way (i.e. having dimensions of (observations x features)) but I'm still confused about why we're naming this variable `WTW` (which, to my mind, strongly suggests `W.T @ W`) when we're computing `W @ W.T`?

Does my question make sense?

Reply:
Because the paper I will write conflicts with sklearn.

Reply:
Sorry, I don't think I'm following you here.  It sounds like you're saying you want to deliberately name a variable contrary to its definition, in a way that looks exactly like an innocent typo / thinko, because that would agree with envisioned notation in an as-of-yet unwritten paper?

I feel like I must be misunderstanding you here-- do you understand my concern about writing `WTW = W @ W.T` and could you help me to understand why it's mistaken, if it is?

Reply:
It reflects the fact that in a program it's easier to have W \in R^{Lxp} but in mathematics have W \in R^{pxL} and I match math papers not code.

Reply:
With the comment at the top of this file explaining the convention, I’m satisfied.